File: android/app/src/main/java/com/github/bfollon/farmaciasdeguardiaensegovia/services/LaLigaBlockingService.kt
/*
 * Copyright (C) 2026  Bruno Follon (@bFollon)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.github.bfollon.farmaciasdeguardiaensegovia.services

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.net.InetAddress
import java.net.URI
import java.util.concurrent.TimeUnit

/**
 * Detects whether our schedule sync server is unreachable because of LaLiga's
 * court-authorized IP blocking (used against Cloudflare-hosted piracy sites during football
 * matches, which collaterally blocks unrelated sites sharing the same IPs — including our
 * own Cloudflare-tunneled server, see [Secrets.scheduleServerBaseUrl]).
 *
 * This is deliberately separate from [NetworkMonitor]: the device can have perfectly good
 * internet while our own server is unreachable, and that combination is what this service
 * distinguishes from a generic offline state (already covered by [ScheduleSyncStatus]).
 *
 * [onServerUnreachable]/[onServerReachable] are called by
 * [PharmacyScheduleRepository]'s `updateSyncStatus` whenever a manifest fetch (our cheapest,
 * most frequent call to our own server) succeeds or fails — see
 * [ScheduleSyncService.fetchManifest].
 */
object LaLigaBlockingService {

    private const val TAG = "LaLigaBlockingService"
    private const val STATUS_URL = "https://hayahora.futbol/estado/blocked-any.txt"
    private const val MIN_RECHECK_INTERVAL_MS = 2 * 60 * 1000L

    private val client = OkHttpClient.Builder()
        .connectTimeout(5, TimeUnit.SECONDS)
        .readTimeout(5, TimeUnit.SECONDS)
        .build()

    @Volatile
    private var lastCheckAtMs = 0L

    @Volatile
    var isLikelyBlocked: Boolean = false
        private set

    /** Call when a request to our own server succeeds — clears any stale blocked state. */
    fun onServerReachable() {
        isLikelyBlocked = false
    }

    /**
     * Call when a request to our own server fails. If the device is online, this checks
     * hayahora.futbol's live list of blocked IPs (rate-limited to once per
     * [MIN_RECHECK_INTERVAL_MS]) to see whether a LaLiga blocking wave is the likely cause.
     */
    suspend fun onServerUnreachable(): Boolean {
        if (!NetworkMonitor.isOnline()) {
            isLikelyBlocked = false
            return false
        }
        val now = System.currentTimeMillis()
        if (now - lastCheckAtMs < MIN_RECHECK_INTERVAL_MS) return isLikelyBlocked
        lastCheckAtMs = now
        isLikelyBlocked = checkBlocking()
        return isLikelyBlocked
    }

    /**
     * Checks whether our own server's IP is among those currently blocked, rather than just
     * whether *some* blocking is active — a resolved-and-matched IP is strong evidence,
     * while an empty/unreachable blocklist is strong evidence against.
     *
     * If our own hostname can't be resolved at all, we have no way to confirm a specific
     * match, so we fall back to correlation (blocklist non-empty) as the best available signal.
     */
    private suspend fun checkBlocking(): Boolean = withContext(Dispatchers.IO) {
        val blockedIps = fetchBlockedIps() ?: return@withContext false
        if (blockedIps.isEmpty()) return@withContext false

        val resolvedIps = resolveServerIps()
        if (resolvedIps.isNullOrEmpty()) {
            DebugConfig.debugWarn("$TAG: could not resolve our own server IP, falling back to correlation")
            true
        } else {
            resolvedIps.any { it in blockedIps }
        }
    }

    private fun fetchBlockedIps(): Set<String>? = try {
        val request = Request.Builder().url(STATUS_URL).build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) return null
            response.body?.string()
                ?.lineSequence()
                ?.map { it.trim() }
                ?.filter { it.isNotEmpty() }
                ?.toSet()
        }
    } catch (e: Exception) {
        DebugConfig.debugError("$TAG: hayahora.futbol check failed", e)
        null
    }

    private fun resolveServerIps(): Set<String>? = try {
        val host = URI(Secrets.scheduleServerBaseUrl).host
        host?.let { InetAddress.getAllByName(it).mapNotNull { addr -> addr.hostAddress }.toSet() }
    } catch (e: Exception) {
        DebugConfig.debugError("$TAG: could not resolve our own server host", e)
        null
    }
}

File: ios/FarmaciasDeGuardiaEnSegovia/Services/LaLigaBlockingService.swift
/*
 * Copyright (C) 2026  Bruno Follon (@bFollon)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import Foundation
#if canImport(Darwin)
import Darwin
#endif

/// Detects whether our schedule sync server is unreachable because of LaLiga's
/// court-authorized IP blocking (used against Cloudflare-hosted piracy sites during football
/// matches, which collaterally blocks unrelated sites sharing the same IPs — including our
/// own Cloudflare-tunneled server).
///
/// This is deliberately separate from `NetworkMonitor`: the device can have perfectly good
/// internet while our own server is unreachable, and that combination is what this service
/// distinguishes from a generic offline state (already covered by `ScheduleSyncStatus`).
///
/// `onServerReachable()`/`onServerUnreachable()` are called from `ScheduleService.updateSyncStatus`
/// whenever a manifest fetch (our cheapest, most frequent call to our own server) succeeds or
/// fails — see `ScheduleSyncService.fetchManifest`.
actor LaLigaBlockingService {
    static let shared = LaLigaBlockingService()

    private let session: URLSession
    private let statusURL = URL(string: "https://hayahora.futbol/estado/blocked-any.txt")!
    private let minRecheckInterval: TimeInterval = 2 * 60

    private var lastCheckAt: Date?
    private(set) var isLikelyBlocked = false

    private init() {
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 5
        config.timeoutIntervalForResource = 5
        session = URLSession(configuration: config)
    }

    /// Call when a request to our own server succeeds — clears any stale blocked state.
    func onServerReachable() {
        isLikelyBlocked = false
    }

    /// Call when a request to our own server fails. If the device is online, this checks
    /// hayahora.futbol's live list of blocked IPs (rate-limited to once per
    /// `minRecheckInterval`) to see whether a LaLiga blocking wave is the likely cause.
    @discardableResult
    func onServerUnreachable() async -> Bool {
        guard NetworkMonitor.shared.isOnline else {
            isLikelyBlocked = false
            return false
        }
        let now = Date()
        if let last = lastCheckAt, now.timeIntervalSince(last) < minRecheckInterval {
            return isLikelyBlocked
        }
        lastCheckAt = now
        isLikelyBlocked = await checkBlocking()
        return isLikelyBlocked
    }

    /// Checks whether our own server's IP is among those currently blocked, rather than just
    /// whether *some* blocking is active — a resolved-and-matched IP is strong evidence,
    /// while an empty/unreachable blocklist is strong evidence against.
    ///
    /// If our own hostname can't be resolved at all, we have no way to confirm a specific
    /// match, so we fall back to correlation (blocklist non-empty) as the best available signal.
    private func checkBlocking() async -> Bool {
        guard let blockedIPs = await fetchBlockedIPs(), !blockedIPs.isEmpty else { return false }

        guard let host = URL(string: Secrets.scheduleServerBaseURL)?.host,
              let resolvedIPs = Self.resolveHostIPs(host), !resolvedIPs.isEmpty
        else {
            DebugConfig.debugPrint("LaLigaBlockingService: could not resolve our own server IP, falling back to correlation")
            return true
        }
        return !resolvedIPs.isDisjoint(with: blockedIPs)
    }

    private func fetchBlockedIPs() async -> Set<String>? {
        do {
            let (data, response) = try await session.data(from: statusURL)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return nil }
            guard let body = String(data: data, encoding: .utf8) else { return nil }
            let ips = body
                .split(whereSeparator: \.isNewline)
                .map { $0.trimmingCharacters(in: .whitespaces) }
                .filter { !$0.isEmpty }
            return Set(ips)
        } catch {
            DebugConfig.debugPrint("LaLigaBlockingService: hayahora.futbol check failed: \(error)")
            return nil
        }
    }

    /// Resolves `host` to its numeric IP addresses via a plain DNS lookup (`getaddrinfo`),
    /// so we can compare against hayahora.futbol's list of currently-blocked IPs.
    private static func resolveHostIPs(_ host: String) -> Set<String>? {
        var hints = addrinfo(
            ai_flags: 0,
            ai_family: AF_UNSPEC,
            ai_socktype: SOCK_STREAM,
            ai_protocol: 0,
            ai_addrlen: 0,
            ai_canonname: nil,
            ai_addr: nil,
            ai_next: nil
        )
        var resultPointer: UnsafeMutablePointer<addrinfo>?
        guard getaddrinfo(host, nil, &hints, &resultPointer) == 0, let firstResult = resultPointer else {
            return nil
        }
        defer { freeaddrinfo(resultPointer) }

        var ips = Set<String>()
        var current: UnsafeMutablePointer<addrinfo>? = firstResult
        while let entry = current {
            var hostBuffer = [CChar](repeating: 0, count: Int(NI_MAXHOST))
            let status = getnameinfo(
                entry.pointee.ai_addr,
                entry.pointee.ai_addrlen,
                &hostBuffer,
                socklen_t(hostBuffer.count),
                nil,
                0,
                NI_NUMERICHOST
            )
            if status == 0 {
                ips.insert(String(cString: hostBuffer))
            }
            current = entry.pointee.ai_next
        }
        return ips.isEmpty ? nil : ips
    }
}
